synapse/handlers/oidc.py

```diff
@@ -957,6 +957,8 @@ async def grandfather_existing_users() -> Optional[str]:
                 # and attempt to match it.
                 attributes = await oidc_response_to_user_attributes(failures=0)
 
+                assert attributes.localpart
```

Comment:
Does anyone know why this is (and the other place) is safe? It's not immediately obvious why this can't be `None` here (it's defined as `Optional[str]`)

Reply:
I suspect in this case it's not. It relies on `_allow_existing_users` not being set at the same time as having a `_user_mapping_provider` which returns `localpart is None`

On the other one, it's impossible to get to `NewUserConsentResource` without having first been through the pick-a-username step which ensures `chosen_localpart` gets set.

Reply:
Thanks!

Oh, and actually in the OIDC case we only use the generated user ID for comparison with existing users, so it doesn't matter if its None? In which case we should instead `return None` if localpart is None?

Reply:
yes.